src/values/set.ts

```diff
@@ -0,0 +1,181 @@
+   * If `after` is not provided, then `data` must be present and represents the
+   * last Page in the Set.
+   */
+  readonly after?: string;
+
+  constructor({ data, after }: { data: T[]; after?: string }) {
+    this.data = data;
+    this.after = after;
+  }
+}
+
+/**
+ * A un-materialize Set. Typically received when a materialized Set contains
+ * another set, the EmbeddedSet does not contain any data to avoid potential
+ * issues such as self-reference and infinite recursion
+ * @see {@link https://fqlx-beta--fauna-docs.netlify.app/fqlx/beta/reference/language/types#set}
+ */
+export class EmbeddedSet {
+  /**
+   * A pagination cursor, used to obtain additional information in the Set.
+   */
+  readonly after: string;
+
+  constructor(after: string) {
+    this.after = after;
+  }
+}
+
+/**
+ * A class to provide an iterable API for fetching multiple pages of data, given
+ * a Fauna Set
+ */
+export class SetIterator<T extends QueryValue>
+  implements AsyncGenerator<T[], void, unknown>
+{
+  readonly #generator: AsyncGenerator<T[], void, unknown>;
+
+  constructor(
+    client: Client,
+    initial: Page<T> | EmbeddedSet | (() => Promise<T>)
+  ) {
+    if (initial instanceof Function) {
+      this.#generator = generateFromThunk(client, initial);
+    } else if (initial instanceof Page || initial instanceof EmbeddedSet) {
+      this.#generator = generatePages(client, initial);
+    } else {
+      throw new TypeError(
+        "Expected 'Pageable<QueryValue> | (() => Promise<T>)', but received " +
+          // @ts-expect-error "Property 'constructor' does not exist on type 'never'."
+          // This is okay, we still want to catch weird inputs from JS
+          initial.constructor.name +
+          " " +
+          JSON.stringify(initial)
+      );
+    }
+  }
+
+  static fromQuery<T extends QueryValue>(
+    client: Client,
+    query: Query
+  ): SetIterator<T> {
+    return new SetIterator<T>(client, async () => {
+      const response = await client.query<T>(query);
+      return response.data;
+    });
+  }
+
+  static fromPageable<T extends QueryValue>(
+    client: Client,
+    pageable: Page<T> | EmbeddedSet
+  ): SetIterator<T> {
+    return new SetIterator<T>(client, pageable);
+  }
+
+  flatten(): FlattenedSetIterator<T> {
+    return new FlattenedSetIterator(this);
+  }
+
+  async next(): Promise<IteratorResult<T[], void>> {
+    return this.#generator.next();
+  }
+
+  async return(): Promise<IteratorResult<T[], void>> {
+    return this.#generator.return();
+  }
+
+  async throw(e: any): Promise<IteratorResult<T[], void>> {
+    return this.#generator.throw(e);
+  }
+
+  [Symbol.asyncIterator]() {
+    return this;
+  }
+}
+
+export class FlattenedSetIterator<T extends QueryValue>
+  implements AsyncGenerator<T, void, unknown>
+{
+  readonly #generator: AsyncGenerator<T, void, unknown>;
+
+  constructor(setIterator: SetIterator<T>) {
+    async function* generateItems<T extends QueryValue>(
```

Comment:
instead of rebuilding this function everytime why not move it out to the broader scope like you did the other generator functions?

Reply:
I started there and then was trying to keep them organized together. But I suppose that's in consistent with the other functions. I'll move this out.